receiver/cloudflarereceiver/logs.go

```diff
@@ -0,0 +1,270 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (l *logsReceiver) processLogs(now pcommon.Timestamp, logs []*models.Log) plog.Logs {
+	pLogs := plog.NewLogs()
+
+	for _, log := range logs {
+		resourceLogs := pLogs.ResourceLogs().AppendEmpty()
+		logRecord := resourceLogs.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+		resourceAttributes := resourceLogs.Resource().Attributes()
+		attrs := logRecord.Attributes()
+
+		logRecord.SetObservedTimestamp(now)
+		resourceAttributes.PutStr("cloudflare.ZoneID", l.cfg.Zone)
+
+		ts := time.Unix(log.EdgeEndTimestamp, 0)
+		logRecord.SetTimestamp(pcommon.NewTimestampFromTime(ts))
+		logRecord.SetSeverityNumber(severityFromAPI(log.EdgeResponseStatus))
+		logRecord.SetSeverityText(severityFromAPI(log.EdgeResponseStatus).String())
+
+		bodyBytes, err := json.Marshal(logs)
+		if err != nil {
+			l.logger.Warn("unable to marshal log into a body string")
+			continue
+		}
+		logRecord.Body().SetStr(string(bodyBytes))
+
+		// These attributes are optional and may not be present, depending on the selected fields.
+		putStringToMapNotNil(attrs, "ClientIP", &log.ClientIP)
+		putStringToMapNotNil(attrs, "ClientRequestHost", &log.ClientRequestHost)
+		putStringToMapNotNil(attrs, "ClientRequestMethod", &log.ClientRequestMethod)
+		putStringToMapNotNil(attrs, "ClientRequestURI", &log.ClientRequestURI)
+		putIntToMapNotNil(attrs, "EdgeEndTimestamp", &log.EdgeEndTimestamp)
+		putIntToMapNotNil(attrs, "EdgeResponseBytes", &log.EdgeResponseBytes)
+		putIntToMapNotNil(attrs, "EdgeResponseStatus", &log.EdgeResponseStatus)
+		putIntToMapNotNil(attrs, "EdgeStartTimestamp", &log.EdgeStartTimestamp)
+		putStringToMapNotNil(attrs, "RayID", &log.RayID)
+	}
+
+	return pLogs
+}
+
+func putStringToMapNotNil(m pcommon.Map, k string, v *string) {
+	if v != nil {
+		m.PutStr(k, *v)
+	}
+}
+
+func putIntToMapNotNil(m pcommon.Map, k string, v *int64) {
+	if v != nil {
+		m.PutInt(k, *v)
+	}
+}
+
+// logRecord wraps a sync Map so it is goroutine safe as well as
+// can have custom marshaling
+type logRecord struct {
+	sync.Mutex
+	LastRecordedTime time.Time `mapstructure:"last_recorded"`
+}
+
+func (r *logRecord) SetLastRecorded(lastUpdated *time.Time) {
+	r.Lock()
+	r.LastRecordedTime = *lastUpdated
+	r.Unlock()
+}
+
+func (l *logsReceiver) syncPersistence(ctx context.Context) error {
+	if l.storageClient == nil {
+		return nil
+	}
+	cBytes, err := l.storageClient.Get(ctx, storageCacheKey)
+	if err != nil || cBytes == nil {
+		l.record = &logRecord{}
+		return nil
+	}
+
+	var cache logRecord
+	if err = json.Unmarshal(cBytes, &cache); err != nil {
+		return fmt.Errorf("unable to decode stored cache: %w", err)
+	}
+	l.record = &cache
+	return nil
+
+}
+
+func (l *logsReceiver) writeCheckpoint(ctx context.Context) error {
+	if l.storageClient == nil {
+		l.logger.Error("unable to write checkpoint since no storage client was found")
+		return errors.New("missing storage client")
```

Comment:
can we validate this before, at start time?

Reply:
added an error check at start time